Gmd performance notebook

mloubout opened the pull request:
This PR adds the performance notebook for reproducibility that performs the convergence analysis and time vs error plots

FabioLuporini commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,433 @@
+    "from examples.seismic.acoustic import AcousticWaveSolver\n",
+    "from examples.seismic import Model, RickerSource, Receiver\n",
+    "from devito import clear_cache\n",
+    "from devito.logger import set_log_level\n",
```
I *think* you can import this from devito directly

mloubout replied:
Can't nope :)

FabioLuporini replied:
Can you add ``set_log_level`` here:
https://github.com/opesci/devito/blob/master/devito/__init__.py#L8
and try again

mloubout replied:
done

FabioLuporini commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,433 @@
+    "# Model with fixed time step value\n",
+    "class ModelBench(Model):\n",
+    "    \"\"\"\n",
+    "    Physical model used for accuracy benchmarking,\n",
```
full stop?

FabioLuporini commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,433 @@
+    "nt = 1501\n",
+    "\n",
+    "# Time axis\n",
+    "\n",
```
blank line?

FabioLuporini commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,433 @@
+   "source": [
+    "# Fine grid model\n",
+    "c0 = 1.500\n",
+    "model = ModelBench(vp=c0, origin=(0., 0.), spacing=(1.0, 1.0), shape=(401, 401), nbpml=40, dtype=np.float64)\n"
```
is float64 really what you want here?

mloubout replied:
Yes, want to get enough numerical precision for the order of errors below.

FabioLuporini commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,433 @@
+    }
+   ],
+   "source": [
+    "solver = AcousticWaveSolver(model, source=src, receiver=rec, time_order=2, space_order=40, dse=None, dle=None)\n",
```
why dse/dle disabled?

mloubout replied:
Performance not important there, but maybe not necessary

mloubout replied:
Also, I want some consistency between `space_orders` to compare timings vs error that dse/dle may mess up with

FabioLuporini replied:
it's just to avoid stupid questions from newbies such as "why the hell are dse/dle set to None here"

mloubout replied:
Looks ok without it, removed

FabioLuporini commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,433 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXEAAAFpCAYAAABqAhhSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzsnX94VPWd719DfjhAiAkECHGAEANm\nWdCoQaNFG1uouuKKLWy1pbe4pX3alb211ueuferdZbf2uT5bbrV77XW3eqtb3cW72JVWu7oLVW6l\nNStRoqZ2JAGmJoSBBBLygwxMYO4fn/nkfOdwzsyZ/AAC5/0885yZ83tmzvd93uf9/Xw ...
```
typo

FabioLuporini replied:
This plot lacks axes names

FabioLuporini commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,433 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXEAAAFpCAYAAABqAhhSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzsnX94VPWd719DfjhAiAkECHGAEANm\nWdCoQaNFG1uouuKKLWy1pbe4pX3alb211ueuferdZbf2uT5bbrV77XW3eqtb3cW72JVWu7oLVW6l\nNStRoqZ2JAGmJoSBBBLygwxMYO4fn/nkfOdwzsyZ/AAC5/0885yZ83tmzvd93uf9/Xw ...
```
can you use the ``e-0.." notation for all yvalues ? And is it possible to decrease the number of precision digits? the yaxis is really verbose ATM

mloubout replied:
Yes, Still digging in opescibench to get nicer plots.

FabioLuporini commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,433 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXEAAAFpCAYAAABqAhhSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzsnX94VPWd719DfjhAiAkECHGAEANm\nWdCoQaNFG1uouuKKLWy1pbe4pX3alb211ueuferdZbf2uT5bbrV77XW3eqtb3cW72JVWu7oLVW6l\nNStRoqZ2JAGmJoSBBBLygwxMYO4fn/nkfOdwzsyZ/AAC5/0885yZ83tmzvd93uf9/Xw ...
```
There's an empty cell at the bottom that can be dropped

mloubout replied:
Oh, yeah allways created when running it, forgot to remove it.

mlange05 commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,415 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXEAAAFpCAYAAABqAhhSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzsvX94lOWd7/8a8sMBQkwgQIgDhBgw\ny4JGDRot2tiGqitWbGGrLT3FLe3VruyptV7fpVc9u+zWXsdry6l2jz3uVk91q7t4Fnuk1a7uQpVT\nac1KlKipHUmAqQkwkEBCfpCBCcz3j8988tzz8Dwzz+QHEHze1zXXM/P8npn7ft/v5/P ...
```
Typo: `reference solution`

mlange05 commented:
Overall I think the approach is sound. I've pointed out a few small things about the general benchmarking infrastructure, but these are mostly nitpicks too.

The plots themselves need a bit more TLC, but this might be done in a follow-up, as it might need some `opescibench` changes as well:
* Error-cost plot: the x-axis tick labels are too dense
* Error analysis: The legend needs to make clear which is reference and which is measured.

Also, the error-cost lines for orders 6 and 8 look a little odd. I wonder if a slightly larger reference/grid would mitigate this? But again, possibly to be addressed in a follow up.

mloubout commented on devito/__init__.py:
```diff
@@ -5,7 +5,7 @@
 from devito.grid import *  # noqa
 from devito.function import Forward, Backward  # noqa
-from devito.logger import error, warning, info  # noqa
+from devito.logger import error, warning, info, set_log_level  # noqa
```
Allow to change it whenever in the notebook at will, I find it easier but can be changed.

mloubout commented on devito/data.py:
```diff
@@ -51,7 +51,7 @@ def __new__(cls, shape, dimensions, halo, dtype):
 
     def __del__(self):
-        if self._c_pointer is None:
+        if getattr(self, '_c_pointer', None) is None:
```
Bug fix @FabioLuporini told me to put there

FabioLuporini replied:
It *is* set in the constructor. But ``__new__`` isn't the only way numpy arrays are created (think about views). It is not that easy to explain in a few lines, but this is all pretty much standard when subclassing numpy arrays. For more info: https://docs.scipy.org/doc/numpy-1.13.0/user/basics.subclassing.html

mlange05 replied:
Ok, fair enough.

mloubout commented on examples/seismic/accuracy.ipynb:
```diff
@@ -0,0 +1,415 @@
+   "outputs": [],
+   "source": [
+    "# Model with fixed time step value\n",
+    "class ModelBench(Model):\n",
```
I find it clearer that way, I also don't really like overwriting the CFL in the normal `Model`  which would lead to people doing stupid shit  (like allways putting 4ms because that's what's in the shot record) and complain it doesn't work.

mlange05 replied:
\<personal opinion> Trying to prevent people from doing "stupid shit" is a futile endeavour as the persistence of idiocy knows no bounds.<\personal opinion> :wink:

But anyway, this might just come down to a matter of taste. I'm personally not a fan because I think this particular use case demonstrates that there is sufficient reason for an override, but I'm happy for it to go in as is.

mloubout commented:
Will fix the typo. As of the plotting, as mentioned above, I am looking into `opescibench` to fix it but shouldn't hold this PR.

Concerning order 6 and 8, I doubt we can get any better without going to insanely small `dt`or a lot of tweaks. Most of what I saw for FD, looks good till order 4 after that it is as good as this.

FabioLuporini commented:
> This notebook is based on the isotropic acoustic wave equation, so perhaps it should live under the acoustic folder?

reminder

mlange05 commented:
> Most of what I saw for FD, looks good till order 4 after that it is as good as this.

Interesting, do you have any references for this?

FabioLuporini commented:
If #472 goes in, please drop the ``Data`` fix in here and use the (more general) one after rebasing on top of that (or just after merging in master, as I presume #472 will land pretty soon)

FabioLuporini commented:
> This notebook is based on the isotropic acoustic wave equation, so perhaps it should live under the acoustic folder?

reminder

mloubout commented:
>reminder

it is

mloubout commented:
@mlange05 Finally got the last piece I wanted, added convergence in time (second order) to justify taking reference solution instead of analytical. I am rewriting the section  accordingly tonight/tmr.
I also changed the way the analytical solution is computed to avoid numerical instabilities (Heaviside and 1/sqrt are dreadfull functions) and I'll add the ref to the paper.

Only thing left is fixing the frickin plotting, I'll keep digging.

mlange05 commented:
Yes, all GTG from me too. The new convergence plots look really nice, although the error-cost plot still looks a bit odd. But we can pick this up later after some digging.

mloubout commented:
Can we merge this

FabioLuporini commented:
Yes, merging...

FabioLuporini commented:
Merged.
